File: examples/python-sdk/ex08_subagent.py
"""ex08 — Multi-agent / subagent delegation. ONE example, two transports:

    python ex08_subagent.py ipc
    python ex08_subagent.py in_process

`jaato.session(mode=...)` picks the transport; the lead (agent="lead") delegates
to subagents (researcher, writer) discovered from .jaato/profiles/. Delegation
spans many turns, so this is the one example that drops to the event API and
waits for SESSION_TERMINATED (not a single turn's completion).

Multi-agent delegation is a harder task than a single ask/stream, so the lead
and its subagents use a model strong enough to orchestrate it reliably
(claude-sonnet-4.5); ex01-07 use the lighter google/gemini-2.5-flash for
single-shot tasks.

The lead must be completion-gated for the wait to resolve: SESSION_TERMINATED
fires only when 'lead' calls signal_completion, which is exposed only when the
profile declares a completion_payload_schema (the `blurb` schema below).
`subagent(preload)` puts the delegation tools on the initial wire, and the
permission policy whitelists them so the spawn is host-independent (not reliant
on a host's ~/.jaato permissions). `socket_path` is ipc-only (ignored
in-process); `env_file` applies to both.

Appears in: pydantic-ai.md §8, langchain.md §8, agno.md §8, claude-agent.md §8,
openai-agents.md §8, strands.md §8.
"""
import asyncio
import sys

import jaato
from jaato_sdk import EventType
from _config import CONN, WORKSPACE, AUTH

mode = sys.argv[1] if len(sys.argv) > 1 else "ipc"


async def main():
    async with jaato.session(
        mode=mode,
        agent="lead",
        workspace_path=WORKSPACE,
        config_root=WORKSPACE + "/.jaato",
        env_file=CONN["env_file"],
        socket_path=CONN["socket_path"],   # ipc-only; ignored in-process
        profile={
            "model": "anthropic/claude-sonnet-4.5", "provider": "openrouter",
            "plugins": ["subagent(preload)", "permission"],
            "plugin_configs": {**AUTH["plugin_configs"],
                               "permission": {"policy": {
                                   "whitelist": {"tools": ["spawn_subagent", "send_to_subagent",
                                                           "close_subagent", "cancel_subagent"]},
                                   "defaultPolicy": "allow"}}},
            "completion_payload_schema": {
                "type": "object", "additionalProperties": False,
                "required": ["blurb"],
                "properties": {"blurb": {"type": "string",
                                         "description": "The final composed blurb."}}},
        },
    ) as s:
        done, out = asyncio.Event(), []
        s.client.subscribe(EventType.AGENT_OUTPUT, lambda e: out.append(getattr(e, "text", "")))
        s.client.subscribe_once(EventType.SESSION_TERMINATED, lambda e: done.set())   # NOT turn.completed
        await s.client.send_message("Research tide pools, then write a blurb from the findings.")
        await done.wait()   # the runtime auto-continues 'lead' as each subagent COMPLETES
        print(f"[{mode}]", "".join(out))


asyncio.run(main())

File: examples/ts-sdk/src/ex08_subagent.ts
// ex08 — Multi-agent / subagent delegation (async, daemon-driven).
//
// Appears in: mastra.md §8. The lead persona delegates via the `subagent`
// plugin; delegation spans many turns, so this drops to the event API and waits
// for SESSION_TERMINATED (not turn.completed).
//
// Doc snippet (verbatim shape):
//
//   await using s = await JaatoClient.session({ url, agent: "lead",
//     profile: { model: "gpt-4o", provider: "openai", plugins: ["subagent"] } });
//   const out: string[] = [];
//   s.client.subscribe(EventTypeValue.AGENT_OUTPUT, (e) => { if (e.text) out.push(e.text); });
//   await new Promise<void>((resolve) => {
//     s.client.subscribeOnce(EventTypeValue.SESSION_TERMINATED, () => resolve());
//     void s.client.sendMessage("Research tide pools, then write a blurb from the findings.");
//   });
//   console.log(out.join(""));
//
// Same deviation as examples/python-sdk/ex08: the prose says the lead must be
// completion-gated, but the doc's inline profile omits the
// completion_payload_schema that exposes signal_completion. This adds the gate (a
// `blurb` schema). Subagent targets (researcher, writer) + the lead persona are
// declarative, so this passes workspacePath + configRoot.
//
// `max_turns` bounds the session so the event-API wait below ALWAYS resolves — a
// completion-gated lead that never calls signal_completion would otherwise hang
// forever. The delegation is fully wired, but whether the lead actually calls
// spawn_subagent is model-dependent; the example demonstrates the wiring either way.
//
// Standing substitutions (see README): `...CONN`, `...AUTH`, the model literal,
// the `plugins` key, + the completion gate.

import { JaatoClient, EventTypeValue } from "@jaato/sdk";
import { CONN, AUTH, WORKSPACE, CONFIG_ROOT } from "./_config.js";

await using s = await JaatoClient.session({
  ...CONN,
  workspacePath: WORKSPACE,
  configRoot: CONFIG_ROOT,
  agent: "lead",
  profile: {
    model: "google/gemini-2.5-flash",
    provider: "openrouter",
    plugins: ["subagent"],
    max_turns: 8,
    completion_payload_schema: {
      type: "object",
      additionalProperties: false,
      required: ["blurb"],
      properties: { blurb: { type: "string", description: "The final composed blurb." } },
    },
    ...AUTH,
  },
});
const out: string[] = [];
s.client.subscribe(EventTypeValue.AGENT_OUTPUT, (e: { text?: string }) => {
  if (e.text) out.push(e.text);
});
await new Promise<void>((resolve) => {
  s.client.subscribeOnce(EventTypeValue.SESSION_TERMINATED, () => resolve()); // NOT turn.completed
  void s.client.sendMessage("Research tide pools, then write a blurb from the findings.");
});
console.log(out.join(""));
